substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,629 @@
+	///
+	/// IMPORTANT: **Checks made in validation need not be repeated here.**
+	fn prepare(
+		self,
+		val: Self::Val,
+		origin: &OriginOf<Call>,
+		call: &Call,
+		info: &DispatchInfoOf<Call>,
+		len: usize,
+		context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError>;
+
+	/// Do any post-flight stuff for an extrinsic.
+	///
+	/// `_pre` contains the output of `prepare`.
+	///
+	/// This gets given the `DispatchResult` `_result` from the extrinsic and can, if desired,
+	/// introduce a `TransactionValidityError`, causing the block to become invalid for including
+	/// it.
+	///
+	/// On success, the caller can return the actual amount of weight consumed by this extension
+	/// during dispatch.
+	///
+	/// WARNING: This function does not automatically keep track of accumulated "actual" weight.
+	/// Unless this weight is handled at the call site, use
+	/// [post_dispatch](TransactionExtension::post_dispatch)
+	/// instead.
+	///
+	/// Parameters:
+	/// - `pre`: `Self::Pre` returned by the result of the `prepare` call prior to dispatch.
+	/// - `info`: Information concerning, and inherent to, the transaction's call.
+	/// - `post_info`: Information concerning the dispatch of the transaction's call.
+	/// - `len`: The total length of the encoded transaction.
+	/// - `result`: The result of the dispatch.
+	/// - `context`: Some opaque mutable context, as yet unused.
+	///
+	/// WARNING: It is dangerous to return an error here. To do so will fundamentally invalidate the
+	/// transaction and any block that it is included in, causing the block author to not be
+	/// compensated for their work in validating the transaction or producing the block so far. It
+	/// can only be used safely when you *know* that the transaction is one that would only be
+	/// introduced by the current block author.
+	fn post_dispatch_details(
+		_pre: Self::Pre,
+		_info: &DispatchInfoOf<Call>,
+		_post_info: &PostDispatchInfoOf<Call>,
+		_len: usize,
+		_result: &DispatchResult,
+		_context: &Context,
+	) -> Result<Option<Weight>, TransactionValidityError> {
+		Ok(None)
+	}
+
+	/// A wrapper for [post_dispatch_details](TransactionExtension::post_dispatch_details) that
+	/// accrues the weight consumed by this extension into the post dispatch information.
+	///
+	/// If `post_dispatch` returns a consumed weight, which should be less than the worst case
+	/// weight provided by [weight](TransactionExtensionBase::weight), that is the value accrued in
+	/// `post_info`.
+	///
+	/// If no weight is returned by `post_dispatch_details`, this function accrues the worst case
+	/// weight.
+	///
+	/// For more information, look into
+	/// [post_dispatch_details](TransactionExtension::post_dispatch_details).
+	fn post_dispatch(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<Call>,
+		post_info: &mut PostDispatchInfoOf<Call>,
+		len: usize,
+		result: &DispatchResult,
+		context: &Context,
+	) -> Result<(), TransactionValidityError> {
```

Comment:
Thinking again, I wonder why `prepare` and `post_dispatch` return a `TransactionValidityError`.
Once the transaction is validated with `validate` it feels more correct to only be able to return a `DispatchError` (or actually `DispatchErrorWithInfo`.

why does the transaction extension should be able to change the validity of a transaction after its dispatch?

And maybe post_dispatch and post_dispatch_details can be merged by just having a signature same as dispatch: i.e. returning a `DispatchResultWithPostInfo`.

I am just giving some thought.